### PR TITLE
mapic: Add event-based nuking on serf

### PR DESCRIPTION
**Pull request by victorges**
Let's fix nuking once and for all. (famous last words)

### CODE REVIEW

File: mapic/mistapiconnector_app.go [hunk 1]
```diff
@@ -506,18 +525,66 @@ func (mc *mac) reconcileLoop(ctx context.Context) {
...
 			glog.Errorf("error executing query on Mist, cannot reconcile err=%v", err)
 			continue
 		}
+		mc.reconcileStreams(mistState)
 		mc.reconcileMultistream(mistState)
 		mc.processStats(mistState)
 	}
 }
 
+func (mc *mac) reconcileStreams(mistState clients.MistState) {
```
Comment (leszko) [inline mapic/mistapiconnector_app.go:541]: I wonder if it's possible to maybe add unit tests for this? Wdyt?

File: mapic/mistapiconnector_app.go [hunk 2]
```diff
@@ -506,18 +525,66 @@ func (mc *mac) reconcileLoop(ctx context.Context) {
...
 		case <-ticker.C:
-		case <-mc.multistreamUpdated:
+		case <-mc.streamUpdated:
 		}
 		mistState, err := mc.mist.GetState()
 		if err != nil {
 			glog.Errorf("error executing query on Mist, cannot reconcile err=%v", err)
 			continue
 		}
+		mc.reconcileStreams(mistState)
```
Comment (leszko) [inline mapic/mistapiconnector_app.go:535]: Nit: Update the doc for the `reconcileLoop(ctx context.Context)`

File: mapic/mistapiconnector_app.go [hunk 3]
```diff
@@ -506,18 +525,66 @@ func (mc *mac) reconcileLoop(ctx context.Context) {
...
 		mc.reconcileMultistream(mistState)
 		mc.processStats(mistState)
 	}
 }
 
+func (mc *mac) reconcileStreams(mistState clients.MistState) {
+	for streamName, _ := range mistState.ActiveStreams {
+		// read map directly to avoid creating streamInfo object even for playback-only (pull) streams
```
Comment (leszko) [inline mapic/mistapiconnector_app.go:543]: I'm not sure I understand this comment. Could you maybe rephrase? We don't create any stream info here 🤔

File: mapic/mistapiconnector_app.go [hunk 4]
```diff
@@ -506,18 +525,66 @@ func (mc *mac) reconcileLoop(ctx context.Context) {
...
+				var err error
+				si, err = mc.refreshStreamInfo(streamName)
+				if err != nil {
+					glog.Errorf("error refreshing stream info streamName=%s err=%v", streamName, err)
+					continue
+				}
+			}
+
+			mc.reconcileSingleStream(si)
+		}
```
Comment (leszko) [inline mapic/mistapiconnector_app.go:559]:
```suggestion
			if mistState.IsIngestStream(streamName) {
			si, err := mc.getStreamInfo(mistStreamName2playbackID(streamName))
			if err != nil {
				glog.Errorf("error getting stream info streamName=%s err=%v", streamName, err)
				continue
			}
			mc.reconcileSingleStream(si)
		}
```

How about this logic? Should work the same and is simpler.

File: mapic/mistapiconnector_app.go (hunk 1 above)
Comment (victorges) [reply mapic/mistapiconnector_app.go:541]: Yeah ideally we would, but I won't have bandwidth for that this week since we're not officially prioritizing this as a task/project. Since this is more of a back-office flow I considered it less of a requirement for now and am willing to merge it as is. Otherwise, if no one is able to pick it up I can come back to this when back from PTO in november.

File: mapic/mistapiconnector_app.go (hunk 3 above)
Comment (victorges) [reply mapic/mistapiconnector_app.go:543]: This reads from the map instead of calling `getStreamInfo`, which would fetch the info object even for the playback streams. I can rephrase it

File: mapic/mistapiconnector_app.go (hunk 4 above)
Comment (victorges) [reply mapic/mistapiconnector_app.go:559]: According to this comment, it is not the same: https://github.com/livepeer/dms-api/blob/835bb9ed3a912bd7ef9d3007b7ef9dc681101abd/mapic/mistapiconnector_app.go#L839

Do you think that case is not that important? I can change to that logic it's much cleaner 😌

File: mapic/mistapiconnector_app.go (hunk 3 above)
Comment (victorges) [reply mapic/mistapiconnector_app.go]: Removed the comment altogether with your suggested change.

File: mapic/mistapiconnector_app.go (hunk 4 above)
Comment (victorges) [reply mapic/mistapiconnector_app.go]: OK I've decided to go forward with it. If the docs I linked are correct, worst case we will wait 30 seconds until the next reconcile loop and it should proceed with the nuke there. We will not be nuking streams that have just started to the point of a "race condition" anywya.

Comment (leszko) [reply mapic/mistapiconnector_app.go]: I think it's correct as I suggested (and as you changed it), because we iterate over Active Streams from Mist (so the scenario mentioned in the comment ☝️ is not important here).

File: mapic/mistapiconnector_app.go (hunk 1 above)
Comment (leszko) [reply mapic/mistapiconnector_app.go:542]: I see. I just checked and adding the unit tests was not so difficult, so added them in 3416b6a9b6d3da4451688367a59ee3d46d0dd9e3

**Review by leszko — APPROVED** [review]
Looks good 👍  Though, keep in mind, I haven't tried if it works!

File: mapic/mistapiconnector_app.go (hunk 1 above)
Comment (victorges) [reply mapic/mistapiconnector_app.go:542]: Awesome, thanks!

**Comment by victorges** [comment]
@leszko It was tested by @iameli-streams yesterday on his daily meta-livestreams 😁 
Thanks again for adding the tests!

File: mapic/mistapiconnector_app_test.go [hunk 5]
```diff
@@ -180,3 +180,84 @@ func TestReconcileMultistream(t *testing.T) {
 	require.ElementsMatch(t, expectedAutoToRemove, recordedAutoRemove)
 	require.ElementsMatch(t, expectedPushToStop, recordedPushStop)
 }
+
+func TestReconcileStreams(t *testing.T) {
```
Comment (victorges) [inline mapic/mistapiconnector_app_test.go:184]: 🎉 ❤️
